store/tikv/client.go

```diff
@@ -172,6 +306,158 @@ func (a *connArray) Close() {
 	close(a.streamTimeout)
 }
 
+type batchCommandsEntry struct {
+	req *tikvpb.BatchCommandsRequest_Request
+	res chan *tikvpb.BatchCommandsResponse_Response
+
+	// Indicated the request is canceled or not.
+	canceled int32
+	err      error
+}
+
+// fetchAllPendingRequests fetches all pending requests from the channel.
+func fetchAllPendingRequests(
+	ch chan *batchCommandsEntry,
+	maxBatchSize int,
+	entries *[]*batchCommandsEntry,
+	requests *[]*tikvpb.BatchCommandsRequest_Request,
+) {
+	// Block on the first element.
+	headEntry := <-ch
+	if headEntry == nil {
+		return
+	}
+	*entries = append(*entries, headEntry)
+	*requests = append(*requests, headEntry.req)
+
+	// This loop is for trying best to collect more requests.
+	for len(*entries) < maxBatchSize {
+		select {
+		case entry := <-ch:
+			if entry == nil {
+				return
+			}
+			*entries = append(*entries, entry)
+			*requests = append(*requests, entry.req)
+		default:
+			return
+		}
+	}
+}
+
+// fetchMorePendingRequests fetches more pending requests from the channel.
+func fetchMorePendingRequests(
+	ch chan *batchCommandsEntry,
+	maxBatchSize int,
+	batchWaitSize int,
+	maxWaitTime time.Duration,
+	entries *[]*batchCommandsEntry,
+	requests *[]*tikvpb.BatchCommandsRequest_Request,
+) {
+	var waitStart, waitEnd time.Time
+	waitStart = time.Now()
+	waitEnd = waitStart
+
+	after := time.After(maxWaitTime)
+Loop1:
+	// Try to collect `batchWaitSize` requests, or wait `maxWaitTime`.
+	for len(*entries) < batchWaitSize {
+		select {
+		case entry := <-ch:
+			if entry == nil {
+				return
+			}
+			*entries = append(*entries, entry)
+			*requests = append(*requests, entry.req)
+		case now := <-after:
+			waitEnd = now
+			break Loop1
+		}
+	}
+
+	if waitEnd == waitStart {
+	Loop2:
+		// Do an additional non-block try.
```

Comment:
Why try twice here?

Reply:
In the first loop the select is blockable, but in the second it's unblockable.